# **Problem Set 2**

*Daniel de Castro*

*APMTH 111*

*September 23, 2021*

# Problem 1

## Part (A)
For a given n-vector $v$,
$$||v||_\infty = max_{1<i<n}|v_i|$$

that is, the absolute value of largest of the components of $v$. In addition, 
$$||v||_2 = \sqrt{\sum_{i=1}^{n}|v_i|^2}$$

Now, let $v_L$ be the component of $v$ with the largest magnitude. Then we can see that $||v||_\infty \leq ||v||_2 \leq \sqrt{n}||v||_\infty$ indeed holds true:
$$||v||_\infty \leq ||v||_2\leq \sqrt{n}||v||_\infty$$
$$|v_L| \leq \sqrt{|v_1|^2 + |v_2|^2 + . . . + |v_L|^2 . . . + |v_n|^2} \leq \sqrt{n}|v_L|$$
$$|v_L|^2 \leq |v_1|^2 + |v_2|^2 + . . . + |v_L|^2 . . . + |v_n|^2 \leq n|v_l|^2$$
$$0 \leq |v_1|^2 + |v_2|^2 + . . . + |v_n|^2 \leq (n-1)|v_l|^2$$
(Note that the middle term in this last inequality is the sum of the squares $n-1$ terms less than or equal to $|v_L|$.)

The vector $v = \begin{bmatrix} 1 \end{bmatrix}$ is a vector in $\mathbb{R}$ for which equality $||v||_\infty = ||v||_2 = \sqrt{n}||v||_\infty$ holds.

## Part (B)
By definition, 
$$||v||_1 = \sum_{i=1}^{n}|v_i|$$
and
$$||v||_2 = \sqrt{\sum_{i=1}^{n}|v_i|^2}$$

Thus, 
$$||v||_2 \leq ||v||_1$$
$$\sqrt{\sum_{i=1}^{n}|v_i|^2} \leq \sum_{i=1}^{n}|v_i|$$
$$\sqrt{|v_1|^2 + |v_2|^2 + . . . + |v_n|^2} \leq |v_1| + |v_2| + . . . + |v_n|$$
$$|v_1|^2 + |v_2|^2 + . . . + |v_n|^2 \leq (|v_1| + |v_2| + . . . + |v_n|)^2$$
$$|v_1|^2 + |v_2|^2 + . . . + |v_n|^2 \leq |v_1|^2 + 2|v_1||v_2| + . . . + 2|v_1||v_n| + |v_2|^2 + 2|v_2||v_3| + . . . + 2|v_{n-1}||v_n| + |v_n|^2$$
$$0 \leq 2|v_1||v_2| + . . . + 2|v_1||v_n| + 2|v_2||v_3| + . . . + 2|v_2||v_n| + 2|v_3||v_4| + . . .$$

The vector $v = \begin{bmatrix} 1 \\ 0 \\ 0 \end{bmatrix}$ is a vector in $\mathbb{R}^3$ for which equality $||v||_2 = ||v||_1$ holds.

## Part (C)
By definition, 
$$||v||_1 = \sum_{i=1}^{n}|v_i|$$
and
$$||v||_\infty = max_{1<i<n}|v_i|$$

Let $v_L$ be the component of $v$ with the largest magnitude. Then we can see that $||v||_1 \leq n||v||_\infty$ indeed holds true:

$$||v||_1 \leq n||v||_\infty$$
$$\sum_{i=1}^{n}|v_i| \leq n \cdot max_{1<i<n}|v_i|$$
$$|v_1| + |v_2| + . . . + |v_L| + . . . + |v_n| \leq n|v_L|$$
$$|v_1| + |v_2| + . . . + |v_n| \leq (n-1)|v_L|$$

Here, the left side of the inequality is the sum of the magnitudes of $n-1$ components, each of which has a magnitude less than or equal to that of $|v_l|$. Clearly, then, the inequality holds true. 

The vector $v = \begin{bmatrix} 1 \\ 1 \\ 1 \end{bmatrix}$ is a vector in $\mathbb{R}^3$ for which equality $||v||_1 = n||v||_\infty$ holds.







(I inserted this line to fix a problem with the document spacing when it is saved as a PDF.)

# Problem 2

## Part A

In [1]:
import numpy as np

# Define A and calculate A_inv
A = np.array([[2, 5, 6], [1, 6, 2],[3, 1, 4]])
A_inv = np.linalg.inv(A)

# Calculate the condition number with the 1-norm

# Create a function to calculate the one-norm of a matrix
def one_norm(B):
  column_sums = np.zeros([len(B[0])])
  for row in B:
    for j in range(len(row)):
      column_sums[j] += np.abs(row[j])
  
  return np.max(column_sums)

print("one-norm conditioning number: ", "{:.3f}".format(one_norm(A) * one_norm(A_inv)))

# Calculate the condition number with the 2-norm 

# Create a function to calculate the infinity-norm of a matrix
def inf_norm(A):
  row_sums = np.zeros(len(A))
  for i in range(len(A)):
    for cell in A[i]:
      row_sums[i] += np.abs(cell)
  
  return np.max(row_sums)

print("infinity-norm conditioning number: ", "{:.3f}".format(inf_norm(A) * inf_norm(A_inv)))

# Calculate the condition number with Frobenius' norm

# Create a function to calculate Frobenius' norm
def fro_norm(A):
  sum = 0
  for row in A:
    for cell in row:
      sum += cell**2
  
  return np.sqrt(sum)

print("Frobenius' norm conditioning number: ", "{:.3f}".format(fro_norm(A) * fro_norm(A_inv)))

one-norm conditioning number:  10.250
infinity-norm conditioning number:  16.792
Frobenius' norm conditioning number:  10.626


With the code above, I calculated the conditioning numbers to be as follows: 

1-norm: K(A) = 10.25

infinite-norm: K(A) = 16.792

Frobenius' norm: K(A) = 10.626

## Part B

In [8]:
import numpy as np

# Define A and calculate A_inv
A = np.array([[2, 5, 6], [1, 6, 2],[3, 1, 4]])
A_inv = np.linalg.inv(A)

# Part i 
one_norm = np.linalg.norm(A, ord=1)
print("one-norm of A:", one_norm)

inf_norm = np.linalg.norm(A, ord=np.inf)
print("infinite-norm of A:", inf_norm)

fro_norm = np.linalg.norm(A, ord='fro')
print("Frobenius' norm of A:", "{:.3f}".format(fro_norm))

# Part ii
print("one-norm conditioning number of A:", "{:.3f}".format(np.linalg.cond(A, p=1)))
print("infinity-norm conditioning number of A:", "{:.3f}".format(np.linalg.cond(A, p=np.inf)))
print("Frobenius' norm conditioning number of A:", "{:.3f}".format(np.linalg.cond(A, p='fro')))

one-norm of A: 12.0
infinite-norm of A: 13.0
Frobenius' norm of A: 11.489
one-norm conditioning number of A: 10.250
infinity-norm conditioning number of A: 16.792
Frobenius' norm conditioning number of A: 10.626


With the code above, I obtained the following results: 

one-norm of A: 12.0

infinite-norm of A: 13.0

Frobenius' norm of A: 11.489

one-norm conditioning number of A: 10.250

infinity-norm conditioning number of A: 16.792

Frobenius' norm conditioning number of A: 10.626

# Problem 3

## Part A

1-norm $K(A) = 3$
        
$\infty$-norm $K(A) = 3$

(See calculations below.)

In [9]:
import numpy as np

# Define A and calculate A_inv
A = np.array([[1, -1], [2, 2]])
A_inv = np.linalg.inv(A)

# 1-norm calculations
print("1-norm K(A) =", "{:.3f}".format(np.linalg.cond(A, p=1)))

# inf-norm calculations
print("infinite-norm K(A) =", "{:.3f}".format(np.linalg.cond(A, p=np.inf)))

1-norm K(A) = 3.000
infinite-norm K(A) = 3.000


## Part (B)
Given an $m \times n$ matrix $A$, the 1-norm and $\infty$-norm are defined as
$$||A||_1 = max_{1<j<n}\sum_{i=1}^{m}|A_{ij}|$$
and
$$||A||_\infty = max_{1<i<m}\sum_{j=1}^{n}|A_{ij}|$$

Let $A$ be the following $2\times2$ matrix
$$A = \begin{bmatrix} a & b \\ c & d \end{bmatrix}$$

and let $A^{-1}$ be its inverse. $A^{-1}$ can be written as
$$A^{-1} = \frac{1}{det(A)} \begin{bmatrix} d & -b \\ -c & a \end{bmatrix} = \frac{1}{ad - bc} \begin{bmatrix} d & -b \\ -c & a \end{bmatrix}$$

Now, let's calculate the condition number $K(A)$ with the 1-norm. 
$$K_1(A) = ||A||_1 \cdot ||A^{-1}||_1$$
$$K_1(A) = max_{1<j<2}\sum_{i=1}^{2}|A_{ij}| \cdot max_{1<j<2}\sum_{i=1}^{2}|A^{-1}_{ij}|$$
$$K_1(A) = max\{a + c, b + d\} \cdot max\{\frac{c + d}{ad - bc}, \frac{a + b}{ad - bc}\}$$
$$K_1(A) = max\{a + c, b + d\} \cdot max\{a + b, c + d\} \cdot \frac{1}{ad - bc}$$

In doing the same for the $\infty$-norm, we find
$$K_\infty(A) = ||A||_\infty \cdot ||A^{-1}||_\infty$$
$$K_\infty(A) = max_{1<i<2}\sum_{j=1}^{2}|A_{ij}| \cdot max_{1<i<2}\sum_{j=1}^{2}|A^{-1}_{ij}|$$
$$K_\infty(A) = max\{a + b, c + d\} \cdot max\{\frac{b + d}{ad - bc}, \frac{a + c}{ad - bc}\}$$
$$K_\infty(A) = max\{a + b, c + d\} \cdot max\{a + c, b + d\} \cdot \frac{1}{ad - bc}$$

Thus, 
$$K_1(A) = K_\infty(A)$$

(Line inserted to fix spacing problem when saving as PDF.)

# Problem 4

## Part (A)
The Frobenius norm of an $m \times n$ matrix $A$ can be written as 
$$||A||_F = \sqrt{ \sum_{i=1}^{m}\sum_{j=1}^{n}|A_{ij}|^2 }$$

We claim that 
$$||A||_F = \sqrt{ \sum_{i=1}^{m}\sum_{j=1}^{n}|A_{ij}|^2 } = \sqrt{ Tr(A^T A) }$$

Consider first the matrix $A^T A$. The components that lie along the main diagonal of $A^T A$ represent the scalar product of one column of $A$ with itself. For example, entry $(A^T A)_{j,j}$ in $A^T A$ would take the following form:
$$(A^T A)_{j,j} = \begin{bmatrix} A_{1,j} & A_{2,j} &  A_{3,j} & . . . & A_{m,j} \end{bmatrix} \begin{bmatrix} A_{1,j} \\ A_{2,j} \\  A_{3,j} \\ . . . \\ A_{m,j}\end{bmatrix}$$
$$(A^T A)_{j,j} = \begin{bmatrix} (A_{1,j})^2 + (A_{2,j})^2 +  (A_{3,j})^2 + . . . + (A_{m,j})^2 \end{bmatrix}$$

$(A^T A)_{j,j}$ thus represents the sum of the squares of the components of one column of $A$. $Tr(A^T A)$, the sum of the main diagonal of $A^T A$, thus represents the sum of the squares of the components for every column of $A$, and can be written as such: 
$$Tr(A^T A) = \sum_{j=1}^{n}[(A_{1,j})^2 + (A_{2,j})^2 +  (A_{3,j})^2 + . . . + (A_{m,j})^2]$$
$$Tr(A^T A) = \sum_{j=1}^{n} \sum_{i=1}^{m} (A_{i,j})^2 = \sum_{i=1}^{m} \sum_{j=1}^{n} (A_{i,j})^2$$
$$\sqrt{Tr(A^T A)} = \sqrt{\sum_{i=1}^{m} \sum_{j=1}^{n} (A_{i,j})^2} = ||A||_F$$

## Part (B)
In part (a), we found that the Frobenius norm can be written as 
$$||A||_F = \sqrt{Tr(A^T A)}$$

Thus, 
$$||UA||_F = \sqrt{Tr((UA)^T (UA))}$$

We can manipulate this expression using the given properties of the Trace of a matrix as follows: 
$$||UA||_F = \sqrt{Tr(A^T U^T U A)}$$
$$||UA||_F = \sqrt{Tr(A^T I A)}$$
$$||UA||_F = \sqrt{Tr(A^T A)}$$
$$||UA||_F = ||A||_F$$

Similar manipulations can be done to find $||AV||_F$:
$$||AV||_F = \sqrt{Tr((AV)^T (AV)}$$
$$||AV||_F = \sqrt{Tr(V^T A^T A V)}$$
$$||AV||_F = \sqrt{Tr(V^T V A^T A)}$$
$$||AV||_F = \sqrt{Tr(I A^T A)}$$
$$||AV||_F = \sqrt{Tr(A^T A)}$$
$$||AV||_F = ||A||_F$$

## Part (C)

### Part (i)
Given that $A = U \Sigma V^T$, where $U$ is an orthogonal $m x m$ matrix, $V$ is an orthogonal $n x n$ matrix, and $\Sigma$ is a diagonal matrix. Thus,
$$||A||_F = \sqrt{Tr((U \Sigma V^T)^T (U \Sigma V^T))}$$
$$||A||_F = \sqrt{Tr(V \Sigma U^T U \Sigma V^T)}$$
$$||A||_F = \sqrt{Tr(V V^T U^T U \Sigma \Sigma)}$$
$$||A||_F = \sqrt{Tr(I^2 \Sigma^2)}$$
$$||A||_F = \sqrt{Tr(\Sigma^2)}$$
$$||A||_F = \sqrt{\sigma_1^2 + \sigma_2^2 + \sigma_3^2 + . . . + \sigma_r^2}$$

### Part (ii)
$$\sigma_{max}(A) \leq ||A||_F \leq \sqrt{r}\sigma_{max}(A)$$
            
In part (i), we showed that 
$$||A||_F = \sqrt{\sigma_1^2 + \sigma_2^2 + \sigma_3^2 + . . . + \sigma_r^2}$$

We can then expand and simplify the inequality we started with as follows: 
$$\sigma_{max}(A) \leq \sqrt{\sigma_1^2 + \sigma_2^2 + \sigma_3^2 + . . . + \sigma_r^2} \leq \sqrt{r}\sigma_{max}(A)$$
$$\sigma_{max}(A)^2 \leq \sigma_1^2 + \sigma_2^2 + \sigma_3^2 + . . . + \sigma_r^2 \leq r\sigma_{max}(A)^2$$

Without loss of generality, let $\sigma_{max}(A) = \sigma_1$. Then
$$\sigma_1^2 \leq \sigma_1^2 + \sigma_2^2 + \sigma_3^2 + . . . + \sigma_r^2 \leq r\sigma_1^2$$
$$0 \leq \sigma_2^2 + \sigma_3^2 + . . . + \sigma_r^2 \leq (r-1)\sigma_1^2$$

Clearly, the middle expression of this inequality is less than or equal to the right, as it consists of $r-1$ terms, each of which is less than or equal to $\sigma_1^2$. The inequality with which we started thus holds true. 

# Problem 5

Given the problem description, we define $||A||_2$ as follows: 
$$||A||_2 = max_{v \neq 0} \frac{||Av||_2}{||v||_2}$$

If $v \in \mathbb{R}^n$, then 
$$v^T v = \sum_{i=1}^n v_i^2$$

Given the definition of $||v||_2$, 
$$||v||_2 = \sqrt{\sum_{i=1}^{n}|v_i|^2}$$

we find that 
$$||v||_2 = v^T v$$

If $A \in \mathbb{R}^{m \times n}$, then $Av \in \mathbb{R}^m$, and we find that 
$$||A||_2 = max_{v \neq 0} \frac{\sqrt{(Av)^T Av}}{\sqrt{v^T v}} = max_{v \neq 0} \frac{\sqrt{v^T A^T Av}}{
\sqrt{v^T v}}$$ 

$$||A||_2 = max_{v \neq 0} \sqrt{\frac{v^T A^T Av}{v^T v}} = \sqrt{max_{v \neq 0} (\frac{v^T A^T Av}{v^T v})}$$

The expression
$$\frac{v^T A^T Av}{v^T v}$$

is a Rayleigh Quotient of $A^T A$; thus,
$$max_{v \neq 0} (\frac{v^T A^T Av}{v^T v})$$

selects for the largest Rayleigh Quotient of $A^T A$, which is also the largest eigenvalue of $A^T A$. We can therefore rewrite $||A||_2$ as 
$$||A||_2 = \sqrt{\lambda_{max}}$$

where $\lambda_{max}$ is the largest eigenvalue of the matrix $A^T A$. 

# References

I referenced the following sources in completing this problem set: 

[Chen, Guangliang. "Lecture 4: Rayleigh Quotients," San José State University](https://www.sjsu.edu/faculty/guangliang.chen/Math253S20/lec4RayleighQuotient.pdf)

[Dahleh, Mohammed et al. "Lectures on Dynamic Systems and Control," Department of Electrical Engineering and Computer Science Massachuasetts Institute of Technology](https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-241j-dynamic-systems-and-control-spring-2011/readings/MIT6_241JS11_chap04.pdf)

["Diagonalization of Symmetric Matrices," University of Illinois Math Department](https://faculty.math.illinois.edu/~franklan/Math415_Diagonalization.pdf)

[Zaky, Zeyad. "Rayleigh Quotient in 3 minutes!"](https://www.youtube.com/watch?v=zhLyPJvcTIs&ab_channel=ZeyadZaky)